File: grok/grok-signup-tool-v2/email_service.js
/**
 * Email Service Module - UPDATED for new tinyhost.shop API
 * Handles temporary email generation and OTP retrieval
 */

const axios = require('axios');

const EMAIL_API_BASE = 'https://tinyhost.shop/api';
const MAX_RETRIES = 12; // 12 attempts x 5s = 60s max wait
const RETRY_DELAY = 5000; // 5 seconds between retries

// Random English name lists for natural-looking emails
const FIRST_NAMES = [
    'james', 'john', 'robert', 'michael', 'david', 'william', 'richard', 'joseph', 'thomas', 'charles',
    'daniel', 'matthew', 'anthony', 'mark', 'steven', 'paul', 'andrew', 'brian', 'kevin', 'jason',
    'mary', 'patricia', 'jennifer', 'linda', 'barbara', 'elizabeth', 'susan', 'jessica', 'sarah', 'karen',
    'emily', 'emma', 'olivia', 'sophia', 'isabella', 'mia', 'charlotte', 'amelia', 'harper', 'evelyn',
    'nathan', 'ryan', 'jacob', 'ethan', 'logan', 'lucas', 'mason', 'alexander', 'henry', 'benjamin',
    'hannah', 'madison', 'chloe', 'grace', 'zoey', 'lily', 'ella', 'riley', 'aria', 'scarlett'
];

const LAST_NAMES = [
    'smith', 'johnson', 'williams', 'brown', 'jones', 'garcia', 'miller', 'davis', 'rodriguez', 'martinez',
    'hernandez', 'lopez', 'gonzalez', 'wilson', 'anderson', 'thomas', 'taylor', 'moore', 'jackson', 'martin',
    'lee', 'perez', 'thompson', 'white', 'harris', 'sanchez', 'clark', 'ramirez', 'lewis', 'robinson',
    'walker', 'young', 'allen', 'king', 'wright', 'scott', 'torres', 'nguyen', 'hill', 'flores',
    'green', 'adams', 'nelson', 'baker', 'hall', 'rivera', 'campbell', 'mitchell', 'carter', 'roberts'
];

/**
 * Generate a random natural-looking username from English names
 * Patterns: firstname.lastname23, firstnamelastname99, f.lastname42, firstname_l56, etc.
 */
function generateNaturalUsername() {
    const first = FIRST_NAMES[Math.floor(Math.random() * FIRST_NAMES.length)];
    const last = LAST_NAMES[Math.floor(Math.random() * LAST_NAMES.length)];
    const num = Math.floor(Math.random() * 999) + 1; // 1-999
    const year = Math.floor(Math.random() * 26) + 85; // 85-10 (birth year style)

    // Randomly pick a pattern
    const patterns = [
        `${first}.${last}${num}`,                          // john.smith42
        `${first}${last}${num}`,                            // johnsmith42
        `${first}_${last}${num}`,                           // john_smith42
        `${first}.${last[0]}${num}`,                        // john.s42
        `${first[0]}${last}${num}`,                         // jsmith42
        `${first}${last}${year}`,                           // johnsmith95
        `${first}.${last}`,                                  // john.smith
        `${first}${num}`,                                    // john42
        `${first}_${last[0]}${num}`,                        // john_s42
        `${first[0]}.${last}${num}`,                        // j.smith42
        `${first}${last.substring(0, 3)}${num}`,            // johnsmi42
        `${first}.${last}${Math.floor(Math.random() * 9)}${Math.floor(Math.random() * 9)}`, // john.smith73
    ];

    return patterns[Math.floor(Math.random() * patterns.length)];
}

/**
 * Generate a temporary email address
 * @returns {Promise<{email: string, domain: string, user: string}>}
 */
async function generateEmail() {
    try {
        console.log('📧 Generating temporary email...');

        // Get random domain from API
        const response = await axios.get(`${EMAIL_API_BASE}/random-domains/`, {
            params: { limit: 5 }
        });

        if (!response.data || !response.data.domains || response.data.domains.length === 0) {
            throw new Error('No domains available from API');
        }

        // Pick random domain
        const domains = response.data.domains;
        const domain = domains[Math.floor(Math.random() * domains.length)];

        // Generate natural-looking username (English names)
        const username = generateNaturalUsername();
        const email = `${username}@${domain}`;

        console.log(`✅ Generated email: ${email}`);
        return { email, domain, user: username };

    } catch (error) {
        console.error('❌ Email generation failed:', error.message);
        throw new Error(`Failed to generate email: ${error.message}`);
    }
}

/**
 * Check email inbox for OTP verification code
 * @param {string} domain - Email domain
 * @param {string} user - Username (without @domain)
 * @returns {Promise<string|null>} OTP code if found, null otherwise
 */
async function checkEmailForCode(domain, user) {
    try {
        console.log(`📬 Checking inbox for ${user}@${domain}...`);

        for (let attempt = 1; attempt <= MAX_RETRIES; attempt++) {
            console.log(`   Attempt ${attempt}/${MAX_RETRIES}...`);

            try {
                // Fetch inbox using new API format
                const response = await axios.get(`${EMAIL_API_BASE}/email/${domain}/${user}/`, {
                    params: {
                        page: 1,
                        limit: 20
                    },
                    timeout: 10000
                });

                const data = response.data;

                if (!data || !data.emails || data.emails.length === 0) {
                    console.log('   No emails yet...');
                } else {
                    console.log(`   Found ${data.emails.length} email(s)`);

                    // Check each email for verification code
                    for (const mail of data.emails) {
                        const code = extractCodeFromEmail(mail);
                        if (code) {
                            console.log(`✅ Found OTP code: ${code}`);
                            return code;
                        }
                    }
                }

            } catch (fetchError) {
                if (fetchError.response && fetchError.response.status === 404) {
                    console.log('   Inbox not created yet...');
                } else {
                    console.log(`   Fetch error: ${fetchError.message}`);
                }
            }

            // Wait before next attempt (except on last attempt)
            if (attempt < MAX_RETRIES) {
                await sleep(RETRY_DELAY);
            }
        }

        console.log('❌ No verification code found after all attempts');
        return null;

    } catch (error) {
        console.error('❌ Email check failed:', error.message);
        return null;
    }
}

/**
 * Extract OTP code from email content
 * @param {Object} email - Email object with subject/body/html_body
 * @returns {string|null} Extracted code or null
 */
function extractCodeFromEmail(email) {
    const content = `${email.subject || ''} ${email.body || ''} ${email.html_body || ''}`;

    // Pattern 1: XXX-XXX format (e.g., OMK-QZN)
    const pattern1 = /\b([A-Z0-9]{3}-[A-Z0-9]{3})\b/i;
    const match1 = content.match(pattern1);
    if (match1) {
        return match1[1].replace('-', ''); // Remove dash to get XXXXXX
    }

    // Pattern 2: 6-digit alphanumeric code
    const pattern2 = /\b([A-Z0-9]{6})\b/i;
    const match2 = content.match(pattern2);
    if (match2) {
        return match2[1];
    }

    // Pattern 3: "code: XXXXXX" or "code is XXXXXX"
    const pattern3 = /code[:\s]+([A-Z0-9]{6})/i;
    const match3 = content.match(pattern3);
    if (match3) {
        return match3[1];
    }

    // Pattern 4: "verification code" followed by code
    const pattern4 = /verification\s+code[:\s]+([A-Z0-9]{6})/i;
    const match4 = content.match(pattern4);
    if (match4) {
        return match4[1];
    }

    return null;
}

/**
 * Sleep helper function
 * @param {number} ms - Milliseconds to sleep
 * @returns {Promise<void>}
 */
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

module.exports = {
    generateEmail,
    checkEmailForCode
};

File: grok/grok-signup-tool/utils/email_service.py
"""Email service for generating temporary emails and retrieving verification codes"""

import httpx
import asyncio
import random
import string
from typing import Optional
from config import (
    TINYHOST_DOMAINS_API,
    BLOCKED_DOMAINS,
    EMAIL_USERNAME_LENGTH,
    EMAIL_CHECK_MAX_RETRIES,
    EMAIL_CHECK_INTERVAL
)
from utils.logger import log_info, log_error, log_success


async def get_random_domain() -> str:
    """
    Fetch random email domain from tinyhost.shop API
    
    Returns:
        str: A random email domain
        
    Raises:
        Exception: If unable to fetch domains
    """
    try:
        log_info("🌐 Fetching random email domains...")
        async with httpx.AsyncClient() as client:
            response = await client.get(TINYHOST_DOMAINS_API, timeout=10.0)
            response.raise_for_status()
            
            data = response.json()
            domains = data.get('domains', [])
            
            if not domains:
                raise Exception("No domains returned from API")
            
            # Filter out blocked domains
            filtered_domains = [
                domain for domain in domains
                if not any(blocked in domain for blocked in BLOCKED_DOMAINS)
            ]
            
            if not filtered_domains:
                raise Exception("No valid domains after filtering")
            
            # Pick a random domain
            domain = random.choice(filtered_domains)
            log_success(f"✅ Selected domain: {domain}")
            return domain
            
    except Exception as e:
        log_error(f"❌ Failed to fetch domains: {str(e)}")
        raise


def generate_random_username(length: int = EMAIL_USERNAME_LENGTH) -> str:
    """
    Generate random alphanumeric username
    
    Args:
        length: Length of username (default from config)
        
    Returns:
        str: Random username
    """
    chars = string.ascii_lowercase + string.digits
    return ''.join(random.choice(chars) for _ in range(length))


async def generate_email() -> str:
    """
    Generate a random temporary email address
    
    Returns:
        str: Complete email address (username@domain.com)
    """
    domain = await get_random_domain()
    username = generate_random_username()
    email = f"{username}@{domain}"
    
    log_info(f"📧 Generated email: {email}")
    return email


async def check_email_for_code(email: str, max_retries: int = EMAIL_CHECK_MAX_RETRIES) -> Optional[str]:
    """
    Check email inbox for verification code using Tinyhost API
    
    Args:
        email: The email address to check
        max_retries: Maximum number of retries (default from config)
        
    Returns:
        str: Verification code (e.g., "OMK-QZN") or None if not found
    """
    log_info(f"📬 Checking email {email} for verification code...")
    
    # Extract username and domain
    username, domain = email.split('@')
    
    # Tinyhost inbox URL format: https://tinyhost.shop/{email}
    inbox_url = f"https://tinyhost.shop/{email}"
    
    for attempt in range(1, max_retries + 1):
        try:
            log_info(f"📨 Attempt {attempt}/{max_retries} to fetch code...")
            
            async with httpx.AsyncClient(timeout=15.0) as client:
                # Get inbox page
                response = await client.get(inbox_url)
                response.raise_for_status()
                
                html_content = response.text
                
                # Look for verification code patterns in HTML
                # Grok sends codes like "OMK-QZN" (3 letters - hyphen - 3 letters/numbers)
                import re
                
                # Pattern 1: Standard code format XXX-XXX
                code_patterns = [
                    r'\b([A-Z]{3}-[A-Z0-9]{3})\b',  # Like OMK-QZN
                    r'\b([A-Z0-9]{6})\b',            # Like UCOX8L (6 chars)
                    r'code[:\s]+([A-Z0-9]{3}-[A-Z0-9]{3})',  # "code: XXX-XXX"
                    r'code[:\s]+([A-Z0-9]{6})',      # "code: XXXXXX"
                ]
                
                for pattern in code_patterns:
                    matches = re.findall(pattern, html_content, re.IGNORECASE)
                    if matches:
                        # Get the first match
                        code = matches[0].upper()
                        log_success(f"✅ Found verification code: {code}")
                        return code
                
                # If no code found, wait and retry
                log_info(f"⏳ No code found yet, waiting {EMAIL_CHECK_INTERVAL}s...")
                await asyncio.sleep(EMAIL_CHECK_INTERVAL)
                
        except httpx.HTTPStatusError as e:
            log_error(f"⚠️ HTTP error {e.response.status_code} (attempt {attempt})")
            await asyncio.sleep(EMAIL_CHECK_INTERVAL)
        except Exception as e:
            log_error(f"⚠️ Error checking email (attempt {attempt}): {str(e)}")
            await asyncio.sleep(EMAIL_CHECK_INTERVAL)
    
    log_error(f"❌ Failed to retrieve code after {max_retries} attempts")
    return None


# For testing purposes
if __name__ == "__main__":
    async def test():
        # Test email generation
        email = await generate_email()
        print(f"Generated email: {email}")
        
        # Test code checking (will fail without implementation)
        # code = await check_email_for_code(email)
        # print(f"Retrieved code: {code}")
    
    asyncio.run(test())
